doc/designs/quic-design/quic-fault-injector.md

```diff
@@ -0,0 +1,507 @@
+hashed. These callbacks will be internal only (i.e. not part of the public API)
+and so only usable by the Fault Injector.
+
+The new libssl callbacks will be as follows:
+
+```` C
+typedef int (*ossl_statem_mutate_handshake_cb)(unsigned char *msgin,
+                                               size_t inlen,
+                                               unsigned char **msgout,
+                                               size_t *outlen,
+                                               void *arg);
+
+typedef void (*ossl_statem_finish_mutate_handshake_cb)(void *arg);
+
+int ossl_statem_set_mutator(SSL *s,
+                            ossl_statem_mutate_handshake_cb mutate_handshake_cb,
+                            ossl_statem_finish_mutate_handshake_cb finish_mutate_handshake_cb,
+                            void *mutatearg);
+````
+
+The two callbacks are set via a single internal function call
+`ossl_statem_set_mutator`. The mutator callback `mutate_handshake_cb` will be
+called after a handshake message has been constructed and is ready to send, but
+before it has been passed through the handshake hashing code. It will be passed
+a pointer to the constructed handshake message in `msgin` along with its
+associated length in `inlen`. The mutator will construct a replacement handshake
+message (typically by copying the input message and modifying it) and store it
+in a newly allocated buffer. A pointer to the new buffer will be passed back
+in `*msgout` and its length will be stored in `*outlen`.
+
+Once libssl has finished using the mutated buffer it will call the
+`finish_mutate_handshake_cb` callback which can then release the buffer and
+perform any other cleanup as required.
+
+### QUIC Pre-Encryption Packets
+
+QUIC Packets are the primary mechansim for exchanging protocol data within QUIC.
+Multiple packets may be held within a single datagram, and each packet may
+itself contain multiple frames. A packet gets protected via an AEAD encryption
+algorithm prior to it being sent. Fault Injector based tests may need to inject
+faults into these packets prior to them being encrypted.
+
+An example test might insert an unrecognised frame type into a QUIC packet to
+confirm that an OpenSSL QUIC client handles it appropriately (e.g. by raising a
+protocol error).
+
+The above functionality will be supported by the following two new callbacks
+which will provide the ability to mutate packets before they are encrypted and
+sent. As for the TLS callbacks these will be internal only and not part of the
+public API.
+
+```` C
+typedef int (*ossl_mutate_packet_cb)(const QUIC_PKT_HDR *hdrin,
+                                     const OSSL_QTX_IOVEC *iovecin, size_t numin,
+                                     QUIC_PKT_HDR **hdrout,
+                                     const OSSL_QTX_IOVEC **iovecout,
+                                     size_t *numout,
+                                     void *arg);
```

Comment:
- Describe how mutator callback can choose not to mutate
- Meaning of return value